File: src/MAC/i_system.m
/*****************************************************************************/
/* D2K: A Doom Source Port for the 21st Century                              */
/*                                                                           */
/* Copyright (C) 2014: See COPYRIGHT file                                    */
/*                                                                           */
/* This file is part of D2K.                                                 */
/*                                                                           */
/* D2K is free software: you can redistribute it and/or modify it under the  */
/* terms of the GNU General Public License as published by the Free Software */
/* Foundation, either version 2 of the License, or (at your option) any      */
/* later version.                                                            */
/*                                                                           */
/* D2K is distributed in the hope that it will be useful, but WITHOUT ANY    */
/* WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS */
/* FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more    */
/* details.                                                                  */
/*                                                                           */
/* You should have received a copy of the GNU General Public License along   */
/* with D2K.  If not, see <http://www.gnu.org/licenses/>.                    */
/*                                                                           */
/*****************************************************************************/


/* Mac path finding and WAD selection UI */

#include <stdlib.h>
#include <string.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "doomdef.h"
#include "doomtype.h"
#include "dstrings.h"
#include "d_main.h"
#include "m_fixed.h"
#include "i_system.h"

#import <Foundation/NSArray.h>
#import <Foundation/NSBundle.h>
#import <Foundation/NSPathUtilities.h>
#import <Foundation/NSString.h>
#import <Foundation/NSFileManager.h>

static NSString *libraryDir(void)
{
  return [@"~/Library/Application Support/PrBoom-Plus" stringByExpandingTildeInPath];
}

static char *NSStringToCString(NSString *str)
{
  char *cStr = malloc([str lengthOfBytesUsingEncoding:NSUTF8StringEncoding] + 1);
  strcpy(cStr, [str UTF8String]);
  return cStr;
}

static char *NSStringToCStringStatic(NSString *str)
{
  static char *cStr = 0;
  static size_t cStrLen = 0;

  int len = [str lengthOfBytesUsingEncoding:NSUTF8StringEncoding] + 1;

  if(!cStr)
  {
    cStr = malloc(len);
    cStrLen = len;
  }
  else if(cStrLen < len)
  {
    free(cStr);
    cStr = malloc(len);
    cStrLen = len;
  }

  strcpy(cStr, [str UTF8String]);
  return cStr;
}

static char *macExeDir = 0;
const char *I_DoomExeDir(void)
{
  if(macExeDir)
    return macExeDir;

  NSString *exeDir = libraryDir();
  [[NSFileManager defaultManager] createDirectoryAtPath:exeDir attributes:nil];

  macExeDir = NSStringToCString(exeDir);
  return macExeDir;
}

const char* I_GetTempDir(void)
{
  if(macExeDir)
    return macExeDir;

  NSString *exeDir = libraryDir();
  [[NSFileManager defaultManager] createDirectoryAtPath:exeDir attributes:nil];

  macExeDir = NSStringToCString(exeDir);
  return macExeDir;
}

char *I_FindFileInternal(const char *wf_name, const char *ext, bool isStatic)
{
  NSArray *paths = [NSArray arrayWithObject:libraryDir()];
  paths = [paths arrayByAddingObject: [[NSBundle mainBundle] resourcePath]];
  paths = [paths arrayByAddingObject: @""];

  char *retval = 0;
  int i;
  for(i = 0; !retval && (i < [paths count]); ++i)
  {
    NSString *path = [NSString stringWithFormat:@"%@/%@",
                      [paths objectAtIndex:i],
                      [NSString stringWithUTF8String:wf_name]];


    if([[NSFileManager defaultManager] isReadableFileAtPath:path])
    {
      if(isStatic)
        retval = NSStringToCStringStatic(path);
      else
        retval = NSStringToCString(path);
    }
  }

  return retval;
}

char *I_FindFile(const char *wf_name, const char *ext)
{
  return I_FindFileInternal(wf_name, ext, false);
}

const char *I_FindFile2(const char *wf_name, const char *ext)
{
  return I_FindFileInternal(wf_name, ext, true);
}

/* vi: set et ts=2 sw=2: */

File: src/POSIX/i_system.c
/*****************************************************************************/
/* D2K: A Doom Source Port for the 21st Century                              */
/*                                                                           */
/* Copyright (C) 2014: See COPYRIGHT file                                    */
/*                                                                           */
/* This file is part of D2K.                                                 */
/*                                                                           */
/* D2K is free software: you can redistribute it and/or modify it under the  */
/* terms of the GNU General Public License as published by the Free Software */
/* Foundation, either version 2 of the License, or (at your option) any      */
/* later version.                                                            */
/*                                                                           */
/* D2K is distributed in the hope that it will be useful, but WITHOUT ANY    */
/* WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS */
/* FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more    */
/* details.                                                                  */
/*                                                                           */
/* You should have received a copy of the GNU General Public License along   */
/* with D2K.  If not, see <http://www.gnu.org/licenses/>.                    */
/*                                                                           */
/*****************************************************************************/


#include "z_zone.h"

#include "m_fixed.h"
#include "i_system.h"
#include "doomdef.h"

#ifdef __GNUG__
#pragma implementation "i_system.h"
#endif
#include "i_system.h"

void I_uSleep(unsigned long usecs) {
#ifdef HAVE_USLEEP
  usleep(usecs);
#else
  /* Fall back on select(2) */
  struct timeval tv = { usecs / 1000000, usecs % 1000000 };
  select(0, NULL, NULL, NULL, &tv);
#endif
}

/*
 * CPhipps - believe it or not, it is possible with consecutive calls to
 * gettimeofday to receive times out of order, e.g you query the time twice and
 * the second time is earlier than the first. Cheap'n'cheerful fix here.
 * NOTE: only occurs with bad kernel drivers loaded, e.g. pc speaker drv
 */

static unsigned long lasttimereply;
static unsigned long basetime;

int I_GetTime_RealTime(void) {
  struct timeval tv;
  struct timezone tz;
  unsigned long thistimereply;

  gettimeofday(&tv, &tz);

  thistimereply = (tv.tv_sec * TICRATE + (tv.tv_usec * TICRATE) / 1000000);

  /* Fix for time problem */
  if (!basetime) {
    basetime = thistimereply;
    thistimereply = 0;
  }
  else {
    thistimereply -= basetime;
  }

  if (thistimereply < lasttimereply)
    thistimereply = lasttimereply;

  lasttimereply = thistimereply;

  return lasttimereply;
}

/*
 * I_GetRandomTimeSeed
 *
 * CPhipps - extracted from G_ReloadDefaults because it is O/S based
 */
unsigned long I_GetRandomTimeSeed(void) {
  /* killough 3/26/98: shuffle random seed, use the clock */
  struct timeval tv;
  struct timezone tz;

  gettimeofday(&tv, &tz);

  return (tv.tv_sec * 1000ul + tv.tv_usec / 1000ul);
}

/* cphipps - I_GetVersionString
 * Returns a version string in the given buffer
 */
const char* I_GetVersionString(char *buf, size_t sz) {
  snprintf(buf, sz, "%s v%s (%s)", PACKAGE_NAME, PACKAGE_VERSION, PACKAGE_URL);
  return buf;
}

/* cphipps - I_SigString
 * Returns a string describing a signal number
 */
const char* I_SigString(char* buf, size_t sz, int signum) {
#if HAVE_DECL_SYS_SIGLIST // NSM: automake defines this symbol as 0 or 1
  if (strlen(sys_siglist[signum]) < sz)
    strcpy(buf, sys_siglist[signum]);
  else
#endif
    sprintf(buf, "signal %d", signum);

  return buf;
}

/* vi: set et ts=2 sw=2: */

